WMR soft limit

ggevay opened the pull request:
This PR implements the WMR soft limit in the form that we agreed on with @aalexandrov. (I'll update the [design doc](https://github.com/MaterializeInc/materialize/pull/18538) tomorrow.) The "soft" means that we don't error out when we reach the limit, but just stop iterating, and consider the current state as the final result.

**The default limit is infinite, i.e., no limit.**

We postponed the hard limit (erroring out when reaching the limit), because we now have proper dataflow cancellation for WMR queries. We can still [consider](https://github.com/MaterializeInc/database-issues/issues/5564) a hard limit later (which would be easy to add after this PR).

We moved away from system/session variables, based on feedback on the [design doc](https://github.com/MaterializeInc/materialize/pull/18538) from Surfaces and Frank.

Instead of system/session variables, users can specify the limit using new SQL syntax (**Edit:** I'm in the process of changing this to a more standard options clause (but leave it at the same place)):
```WITH MUTUALLY RECURSIVE MAXITERATIONS 42```
This syntax allows for separate limits for each WMR block, which is [important](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md#queries-with-multiple-wmr-clauses----usability) when a query has multiple WMR blocks.

The first commit is just changing iteration numbers to use `u64` instead of `usize`, based on a discussion in the office hours. The second commit runs `cargo fmt`, which I will squash into the first one, of course. (Just wanted to separate the diff for review.)

The second commit is the meat.

@aalexandrov, is the EXPLAIN format ok? (HIR, MIR, LIR) Note that I'm not currently testing `linear_chains`, because it [doesn't seem to work for WMR (and nobody seems to be using it)](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1682451462451629). We can discuss whether to fix it or deprecate it. (I'll open an issue tomorrow.)



#### Keyword

The keyword is tentative; we can still decide that before merging, but I like `MAXITERATIONS`. (@ggnall) Some possible alternatives:
- `MAXDEPTH 42`: users might think of recursive function calls, which is not what's happening here.
- `LIMIT 42`: It's not clear what are we limiting: iterations, records, time, ... Also, it could be confused with ORDER BYs LIMIT.
- `(ITERATE 42 TIMES)`: This one would also be ok I guess, but I'd vote for simplicity.
- `RECURSIONLIMIT`: Might be ok as well, but I wanted to avoid (internal) confusion with our `pub const RECURSION_LIMIT`, which is for something totally different.
- `MAXRECURSION`: Edit: The MySQL option with the same keyword would correspond to a _hard_ limit, so let's not align our soft limit's keyword with that.
- `ITERATIONS`: ([from Jan](https://github.com/MaterializeInc/materialize/pull/18538#discussion_r1189689052)) We don't have to call it a limit, since running to an exactly specified number of iterations is the same as stopping when either a limit is reached or fixpoint is reached. But I'm a bit worried that some users won't make this extra mental step of realizing that it might execute less iterations if a fixpoint is reached earlier, and might get worried that it will take more steps than necessary.

### Motivation

  * This PR adds a known-desirable feature: https://github.com/MaterializeInc/database-issues/issues/5409

### Tips for reviewer

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated design doc: https://github.com/MaterializeInc/materialize/pull/18538
  - (The design doc currently doesn't reflect the move away from session/system variables as well as the hard limit's importance going down with the dataflow shutdown. I'll update the design doc a bit later.)
  <!-- Reference the design in the description. -->
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
  - We do have a protobuf change for the LIR `LetRec`, but I asked Jan whether we need to be backwards-compatible, and he said "Nope, we are not durably storing serialized IRs anywhere, I’m pretty sure. So no need to be backwards-compatible in the protobufs"
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

mjibson commented on src/sql-parser/src/ast/defs/query.rs:
```diff
@@ -362,10 +362,12 @@ impl<T: AstInfo> AstDisplay for CteBlock<T> {
                     f.write_str("WITH MUTUALLY RECURSIVE ");
-                    if let Some(max_depth) = max_depth {
-                        f.write_node(max_depth);
+                    if let Some(max_iterations) = max_iterations {
```
This needs a test in https://github.com/MaterializeInc/materialize/blob/main/src/sql-parser/tests/testdata/select

ggevay replied:
Thanks! Just saw your msg on Slack. I'll definitely add a test.

(I'll ping you when the PR is ready for review.)

ggevay replied:
Done.

ggevay commented:
Ready for review! @MaterializeInc/surfaces, @aalexandrov, @vmarcos (rendering).

mjibson commented on src/sql-parser/src/ast/defs/query.rs:
```diff
@@ -338,8 +344,11 @@ impl<T: AstInfo> CteBlock<T> {
             }
-            CteBlock::MutuallyRecursive(list) => {
-                for cte in list.iter() {
+            CteBlock::MutuallyRecursive(MutRecBlock {
```
As an aside (don't need to do anything in this PR): having logic in the parser is unfortunate and it'd be nice for it to not be here.

mjibson commented on src/sql-parser/src/parser.rs:
```diff
@@ -4619,17 +4619,31 @@ impl<'a> Parser<'a> {
+                            return parser_err!(
+                                parser,
+                                parser.peek_prev_pos(),
+                                "MAXITERATIONS has to be greater than 0"
```
This logic should live in the planner, not the parser.

ggevay replied:
Ok, I moved it.

ggevay replied:
The PR should be in it's final form now. Could you please do a final check @mjibson?

teskje commented on src/compute/src/render/mod.rs:
```diff
@@ -100,6 +100,7 @@
 //! stream. This reduces the amount of recomputation that must be performed
 //! if/when the errors are retracted.
 
+use itertools::izip;
```
Nit: This should be in the third-party import block below.

ggevay replied:
The import block below is DD/Timely. I created a new import block between std and the DD/Timely block. Is this what you meant?

teskje replied:
Normally (though I don't think that's spelled out anywhere), we structure our imports into blocks like this:

* `std` imports
* third-party imports
* `mz_` imports
* `crate` imports
* `super` imports
* `self` imports

(The last three blocks I'm actually not sure about. I think they are sometimes merged into one.)

According to this, DD, Timely, and `itertools` all should be in the third-party import block. I don't think we usually do dedicated import blocks for DD/Timely. In this case it just looks like we do because those are the only third-party crates we use in this module.

ggevay replied:
Ok, fixed, thanks!

teskje commented on src/compute/src/render/mod.rs:
```diff
@@ -646,6 +658,25 @@ where
                 }
+
+                match max_iter {
+                    Some(max_iter) => {
```
Nit: `if let Some(max_iter) = max_iter { ... }` would be slightly more readable.

ggevay replied:
Yes, thanks, changed it.

philip-stoev reviewed (approved):
Thank you very much for the tests -- I could not figure out any additional ones to add.

The feature held under manual experimentation and I could not get it to become wedged .

The diff looks less formidable with whitespace ignored.

teskje commented on src/compute-client/src/plan.proto:
```diff
@@ -76,6 +76,8 @@ message ProtoPlan {
         repeated mz_expr.id.ProtoLocalId ids = 1;
         repeated ProtoPlan values = 2;
+        repeated uint64 max_iters = 4;
+        repeated bool max_iters_present = 5;
```
I assume there is no `repeated optional uint64` that would make this second field unnecessary?

Relatedly, but not really relevant to this PR, I have been wondering why `LetRec` is represented with separate lists for ids and values (and now max_iters). ISTM that instead having a single `bindings` list with `(id, value, max_iter)` entries would be more convenient since it doesn't require ensuring that the lists have the same length all the time. But I'm probably missing something.

aalexandrov replied:
We discussed that this might be a better representation but decided to not change it as part of the ongoing epic. I think it must be a skunkworks project or something similar, unless we find ~3-4 days of in-band work to do this as part of addressing tech debt.

teskje replied:
Makes sense, thanks!

Regarding `max_iters_present`, Alex mentioned in another comment:
> One general comment: we should absolutely prohibit people setting the MAX_ITERATIONS for some binding to 0, otherwise we run the risk of all sorts of incorrect optimizations for those blocks (or unnecessary complicated transform code).

Doing that would allow us to drop `max_iters_present` here and just use 0 to mean "no limit set".

teskje replied:
I'd prefer it, since it statically removes the possibility that the two lists might become inconsistent (e.g. have different lengths), so there is one thing less to check at runtime. But your argument about special values is valid too, so I don't oppose leaving things as they are.

ggevay replied:
Changed it.

teskje commented on src/expr/src/explain/text.rs:
```diff
@@ -298,15 +298,52 @@ impl MirRelationExpr {
+                assert_eq!(ids.len(), values.len());
+                assert_eq!(ids.len(), max_iters.len());
+                let bindings = itertools::izip!(ids.iter(), values.iter(), max_iters.iter())
+                    .collect::<Vec<(&LocalId, &MirRelationExpr, &Option<u64>)>>(); // CLion needs these explicit types
```
I'm not a huge fan of making our code specific to the limitations of specific IDEs!

I can't think of a good solution, but it makes my wonder why we are not running into this issue more often. Is this the only instance where CLion is not able to able to handle code that rustc is, or do you usually ignore errors? I assume more people are not using CLion, so it must happen from time to time that someone checks in code that's not compatible with it.

ggevay replied:
Yes, there are some situations where it can't figure out some types. Here, I think the `izip` macro trips it up. There are also some situations where it detects a type error and makes a red underlining when in fact there is no type error. But these situations are rare enough that I can just ignore it with negligible impact on my productivity.

I deleted the explicit type annotation.

teskje commented on src/expr/src/explain/text.rs:
```diff
@@ -298,15 +298,52 @@ impl MirRelationExpr {
+                    } else {
+                        &None
+                    }
+                };
```
This looks like it can be a single `reduce`:
```rust
let all_max_iters_same = max_iter.iter().reduce(|first, i| {
    if i == first { first } else { &None }
});
```

ggevay replied:
Thanks!

teskje commented on src/transform/src/normalize_lets.rs:
```diff
@@ -161,6 +161,7 @@ impl NormalizeLets {
     }
 }
 
+use crate::normalize_lets::support::replace_bindings_from_map;
```
Nit: `use support::replace_bindings_from_map`.

ggevay replied:
Fixed

teskje reviewed (approved):
General compute parts LGTM, I just have a couple stylistic comments. I'll defer to @aalexandrov for the transform changes.

vmarcos reviewed (approved):
Rendering changes look good to me; it would be ideal if we'd include test(s) for nested WMR blocks with different limits also in execution (`test/sqllogictest/with_mutually_recursive.slt`), though, not only in planning.

aalexandrov commented on test/sqllogictest/explain/physical_plan_as_text.slt:
```diff
@@ -855,3 +855,51 @@ Source materialize.public.r
+      Get::PassArrangements l1
+        raw=true
+  With Mutually Recursive
+    cte [MaxIterations None] l1 =
```
nit: stylistically parameters of AST nodes have been printed as `max_iterations` elsewhere, so I think
```text
cte [max_iterations=10] l1 =
```
is a bit more consistent with the rest of the format.

ggevay replied:
Yes, fixed, thx!

aalexandrov commented:
One general comment: we should absolutely prohibit people setting the MAX_ITERATIONS for some binding to `0`. There are some `NonZero~` types in https://doc.rust-lang.org/stable/std/num/index.html and I guess the most precise way to enforce this constraint is to use one of those.

ggevay commented on src/compute-client/src/plan/mod.rs:
```diff
@@ -11,6 +11,7 @@
 
 #![warn(missing_debug_implementations)]
 
+use itertools::Itertools;
```
Fixed

ggevay commented on src/compute-client/src/explain/text.rs:
```diff
@@ -20,6 +20,7 @@
 //!    pairs on the same line or as lowercase `$key` fields on indented lines.
 //! 5. A single non-recursive parameter can be written just as `$val`.
 
+use itertools::{izip, Itertools};
```
Fixed

ggevay commented:
Thanks for the reviews! In addition to addressing the inline comments, I made the following changes:

I added the test from @philip-stoev.

I added the test that @vmarcos suggests.

I changed the LIR EXPLAIN to not print it if None, as @aalexandrov suggested.

> we should absolutely prohibit people setting the MAX_ITERATIONS for some binding to 0, otherwise we run the risk of all sorts of incorrect optimizations for those blocks (or unnecessary complicated transform code).

Yes, I'm throwing an error for 0. (while creating the HIR from the AST)

> There are some NonZero~ types

Hmm, nice! I changed the code to use this. (I didn’t properly implement `MzReflect for NonZeroU64`, but I hope that’s ok. We are not planning to use lowertest for WMR I guess, since we are moving away from lowertest anyway. @aalexandrov)

ggevay commented:
@benesch, you were auto-added as a reviewer due to adding a small thing in mz-ore in `src/ore/src/option.rs`.

ggevay commented:
Is the keyword and syntax ok? Who has the authority on this? @ggnall? A simple example:
```
WITH MUTUALLY RECURSIVE MAXITERATIONS 3
  cnt (i int) AS (
    SELECT 1 AS i
    UNION
    SELECT i+1 FROM cnt)
SELECT * FROM cnt;
```

benesch reviewed (changes requested):
Yes, we need a design proposal for the syntax. A short design document would be ideal, but a comment in a GitHub issue or on this PR would do fine.

The current syntax does not follow any of our established idioms for specifying options or naming parameters. We have the start of a guide for these idioms here: https://www.notion.so/materialize/SQL-design-principles-2e2069797f7c46b8b4c7dd10043afdcd

I suggest that we model this after the existing `OPTIONS` and `WITH` clauses:

```sql
WITH MUTUALLY RECURSIVE (MAX ITERATIONS 1)
   cnt (i int) AS (SELECT ...)
```

Where `MAX ITERATIONS` is two separate tokens, per our style guide, and the bit in the parentheses is a standard options block, where you can also say `MAX ITERATIONS = 3`.

Is "iterations" the right term here? Given that the feature is "with mutually **recursive**", would we prefer `RECURSION LIMIT` to `MAX ITERATIONS`?

aalexandrov commented on src/compute-client/src/plan/mod.rs:
```diff
@@ -590,9 +594,22 @@ impl RustType<ProtoPlan> for Plan {
                         ids: ids.into_proto(),
+                        max_iters: max_iters
+                            .into_iter()
+                            .map(|d| match d {
```
minor nit: can you add this 

```rust
impl RustType<u64> for Option<NonZeroU64> {
    fn into_proto(&self) -> u64 {
        match self {
            Some(d) => d.get(),
            None => 0,
        }
    }

    fn from_proto(proto: u64) -> Result<Self, TryFromProtoError> {
        Ok(NonZeroU64::new(proto))
    }
}
```

after this line

```rust
rust_type_id![bool, f32, f64, i32, i64, String, u32, u64, Vec<u8>];
```

and then use `max_iters.into_proto()` and `proto.max_iters.into_rust()?`?

aalexandrov reviewed (approved):
Looks good, from my original suggestions I think the only thing missing is changing the output format of the new attribute from
```text
[MaxIterations $x]
```
to
```text
[max_iterations=$x]
```

ggevay commented:
I've addressed all comments, including changing the syntax based on the SQL design principles notion doc and SELECT's expected group size option. An example for the syntax:
```
WITH MUTUALLY RECURSIVE (ITERATION LIMIT 6)
  cnt (i int) AS (
    (WITH MUTUALLY RECURSIVE (ITERATION LIMIT = 3)
       cnt (i int) AS (
         SELECT 1 AS i
         UNION
         SELECT i+1 FROM cnt)
       SELECT i FROM cnt
    )
    UNION
    SELECT i+100 FROM cnt)
SELECT i FROM cnt;
```
We can decide the exact keywords later, after we finalize the keywords for WITH MUTUALLY RECURSIVE itself. (One option that came up was WITH REPEATEDLY, which doesn't have RECURSIVE in it, so the word ITERATION would be ok for it.)

@benesch, @ggnall, could you please take a quick look at the syntax?

The surfaces code changed because of the different option parsing. @mjibson, could you please check the new parsing and planning?

I also updated the [design doc](https://github.com/MaterializeInc/materialize/pull/18538).

def- commented on src/compute/src/render/mod.rs:
```diff
@@ -271,67 +273,59 @@ pub fn build_compute_dataflow<A: Allocate>(
-                        );
-                    }
+                for (id, (oks, errs)) in imported_sources.into_iter() {
+                    let bundle = crate::render::CollectionBundle::from_collections(
```
This looks like an interesting block to test

ggevay replied:
This is actually already covered by `with_mutually_recursive.slt`.

def- replied:
Thanks, you are right, this was an issue in my coverage implementation, fixed with https://github.com/MaterializeInc/materialize/pull/19193

def- commented on src/compute/src/render/mod.rs:
```diff
@@ -271,67 +273,59 @@ pub fn build_compute_dataflow<A: Allocate>(
-                        tokens.insert(object.id, object_token);
+                // Import declared indexes into the rendering context.
+                for (idx_id, idx) in &dataflow.index_imports {
+                    let export_ids = dataflow.export_ids().collect();
```
Interesting block to test

ggevay replied:
This is indeed not covered, thanks! I'll add a test for this. (Although this PR makes only a trivial change to this part.)

ggevay replied:
Added a test in aebe703f4e3fe72b3982094dc0cb116920148231

def- commented on src/compute/src/render/mod.rs:
```diff
@@ -656,6 +660,22 @@ where
                 }
+
+                if let Some(max_iter) = max_iter {
+                    // We swallow the results of the `max_iter`th iteration, because
```
Interesting block to test

ggevay replied:
Hmm, well, this is definitely covered in `with_mutually_recursive.slt` (by most of the new tests), as this is the core of the PR. I verified this now by putting a crash into this block and running `with_mutually_recursive.slt`.

What tests does coverage take into account? Does it consider slt tests? Or how can it be that it shows it as not covered?

def- replied:
It should consider slt tests: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/code-coverage.md I'm seeing a problem with the coverage data here, will investigate.

def- commented on src/expr/src/explain/text.rs:
```diff
@@ -298,15 +298,42 @@ impl MirRelationExpr {
+
                 if ctx.config.linear_chains {
-                    writeln!(f, "{}With Mutually Recursive", ctx.indent)?;
+                    write!(f, "{}With Mutually Recursive", ctx.indent)?;
```
This entire block is untested it seems?

ggevay replied:
Yes, because linear chains currently doesn't work with WMR, see https://github.com/MaterializeInc/materialize/issues/19012.

def- reviewed (commented):
Some interesting spots from code coverage report marked inline: https://buildkite.com/materialize/coverage/builds/89

mjibson reviewed (commented):
Surfaces parts lgtm

benesch commented:
New syntax looks much better, nice! I signal boosted in #devex (https://materializeinc.slack.com/archives/C015RHB3LDR/p1683779552919369) for additional feedback. "Iteration limit" sounds a little funny to my ear, but no strong feelings.

morsapaes commented:
> We can decide the exact keywords later, after we finalize the keywords for WITH MUTUALLY RECURSIVE itself. (One option that came up was WITH REPEATEDLY, which doesn't have RECURSIVE in it, so the word ITERATION would be ok for it.)

Any user that is familiar with recursion in SQL will have a mental mapping for `WITH RECURSIVE`, so it doesn't sound right to me to use an entirely different term for it (like `WITH REPEATEDLY`). Are we **forced** to use the `MUTUALLY` keyword, or would it be possible to relax this to `WITH RECURSIVE`, and spell out how we depart from the SQL standard implementation?

For the same reason, my preference for the iteration limit would also be for a keyword that is used in other systems, like `MAXRECURSION` or `MAXDEPTH `. AFAIU, these refer to the maximum depth level, though (which seems to be how most databases limit recursive queries?), so if that's a fundamentally wrong way to think about it, it might be safer to use `MAXITERATIONS`.

benesch commented:
> From my point of view, simply using `WITH RECURSIVE` sounds ok, but Frank said that Nikhil might not want that. The only potential issue that I can see is that we have a semantic difference, i.e., some recursive queries have different results between Postgres and Materialize.

Yeah, I want to keep the door open for supporting the SQL standard's semantics for `WITH RECURSIVE`. It may be important for a customer one day.

@petrosagg had the take of: why bother with the `RECURSIVE` keyword at all? Just allow CTEs in a normal `WITH` block to refer to one another. I think that's more plausible, because that's a strict extension to the SQL spec, rather than 

> MySQL errors out when reaching MAXRECURSION, which would correspond to our hard limit. In contrast, the soft limit (this PR) simply produces the current state as the final result when reaching the limit. For this reason, I think we shouldn't align the keyword.

Oh, interesting! That makes sense for debugging (you want to watch it progress), but kind of scary if you were running in production. Should we be very explicit with our keyword and include something like `SOFT` in the name? `SOFT ITERATION LIMIT`, for example? Although I'm not sure it's exactly a "soft limit"—I think of a "soft" vs "hard" as "changeable upon request" vs "enforced", whereas this is about "errors or not if limit reached." Naming is hard!

Just noodling:

```
WITH ... (ITERATION LIMIT = 256, REQUIRE FIXPOINT = false)
WITH ... (ITERATION LIMIT = 256, REQUIRE FIXPOINT = true)
WITH ... (ITERATION LIMIT = 256) -- REQUIRE FIXPOINT defaults to true
```

----

I think to avoid blocking this PR any longer we should move forward with either `ITERATION LIMIT` or `RECURSION LIMIT`, but let's sync again on the syntax holistically once this is closer to stabilization.

teskje commented:
Another idea I had was to just use the keyword `ITERATIONS` without calling it a limit: https://github.com/MaterializeInc/materialize/pull/18538#discussion_r1189689052.

ggevay commented:
I like Petros' and Jan's simplifying ideas:
- Simply `WITH`, no `RECURSIVE` is needed.
- Simply `ITERATIONS`, no `LIMIT` is needed.

I'm not sure about the [soft vs. hard limit](https://github.com/MaterializeInc/materialize/pull/18966#issuecomment-1544785102). I think if the docs are very explicit about not erroring but simply stopping by default, then it's ok to simply stop by default.

Merging now with ITERATION LIMIT, and then let's decide these things as a follow-up, to already let people use the feature (internally), and also to avoid rebasing this PR again and again.
